mixer/cmd/root.go

```diff
@@ -63,36 +99,78 @@ func Execute() {
+
+	initCmd.Flags().BoolVar(&initFlags.all, "all", false, "Create a mix with all Clear bundles included")
+	initCmd.Flags().IntVar(&initFlags.clearver, "clear-version", 1, "Supply the Clear version to compose the mix from")
+	initCmd.Flags().IntVar(&initFlags.mixver, "mix-version", 0, "Supply the Mix version to build")
+	initCmd.Flags().StringVar(&config, "config", "", "Supply a specific builder.conf to use for mixing")
+	initCmd.Flags().StringVar(&initFlags.upstreamurl, "upstream-url", "https://download.clearlinux.org", "Supply an upstream URL to use for mixing")
+
+	externalDeps[initCmd] = []string{
 		"git",
-		"hardlink",
-		"m4",
-		"openssl",
-		"parallel",
-		"rpm",
-		"yum",
 	}
-	for _, dep := range deps {
-		if _, err := exec.LookPath(dep); err != nil {
-			return fmt.Errorf("failed to find program %q: %v", dep, err)
+}
+
+// externalDeps let commands keep track of their external program dependencies. Those will be
+// verified when the command is executed, just make sure it is filled at initialization.
+var externalDeps = make(map[*cobra.Command][]string)
+
+func checkCmdDeps(cmd *cobra.Command) error {
+	var deps []string
+	for ; cmd != nil; cmd = cmd.Parent() {
+		deps = append(deps, externalDeps[cmd]...)
+	}
+	sort.Strings(deps)
+
+	var missing []string
+	for i, dep := range deps {
+		if i > 0 && deps[i] == deps[i-1] {
+			// Skip duplicate.
+			continue
 		}
+		_, err := exec.LookPath(dep)
+		if err != nil {
+			missing = append(missing, dep)
+		}
+	}
+	if len(missing) > 0 {
+		return errors.Errorf("missing following external programs: %s", strings.Join(missing, ", "))
 	}
 	return nil
 }
 
-func init() {
-	if err := checkDeps(); err != nil {
-		fmt.Fprintf(os.Stderr, "Error: %s\n", err)
-		os.Exit(1)
+func checkAllDeps() bool {
+	var allDeps []string
+	for _, deps := range externalDeps {
+		allDeps = append(allDeps, deps...)
 	}
+	sort.Strings(allDeps)
 
-	RootCmd.AddCommand(initCmd)
+	var max int
+	for _, dep := range allDeps {
+		if len(dep) > max {
+			max = len(dep)
+		}
+	}
 
-	initCmd.Flags().BoolVar(&initFlags.all, "all", false, "Create a mix with all Clear bundles included")
-	initCmd.Flags().IntVar(&initFlags.clearver, "clear-version", 1, "Supply the Clear version to compose the mix from")
-	initCmd.Flags().IntVar(&initFlags.mixver, "mix-version", 0, "Supply the Mix version to build")
-	initCmd.Flags().StringVar(&config, "config", "", "Supply a specific builder.conf to use for mixing")
-	initCmd.Flags().StringVar(&initFlags.upstreamurl, "upstream-url", "https://download.clearlinux.org", "Supply an upstream URL to use for mixing")
+	fmt.Println("Programs used by Mixer commands:")
+	ok := true
+	for i, dep := range allDeps {
+		if i > 0 && allDeps[i] == allDeps[i-1] {
+			// Skip duplicate.
+			continue
+		}
+		_, err := exec.LookPath(dep)
+		if err != nil {
+			fmt.Printf("  %-*s not found\n", max, dep)
+			ok = false
+		} else {
+			fmt.Printf("  %-*s ok\n", max, dep)
+		}
+	}
```

Comment:
Both here and in checkCmdDeps (at line 130), you have a "loop through dependencies, skipping duplicates, and call exec.LookPath" block that I probably would have pulled out and made a common thing. But I see you're doing a different thing in the output, and the [Rule of Three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)) says if you need it twice just copy, so I'm fine with it how it is.